rfc/image-discoverability.md

```diff
@@ -0,0 +1,127 @@
+- a commit hash of the Constellation monorepo that is used to build the images (e.g. `cc0de5c68d41f31dd0b284d574f137e0b0ad106b`)
+
+To identify images belonging to one invocation of the build pipeline, the pipeline uses a unique identifier for the set of images, referred to as `image version uid` that combines the version number and the commit hash (e.g. `v2.2.0-cc0de5c68d41f31dd0b284d574f137e0b0ad106b`).
+
+The build pipeline produces as outputs:
+
+- a raw OS image for every CSP
+- a set of measurements for each raw OS image
+- one or more images uploaded to each CSP (e.g. AWS AMIs, Azure images, GCP images)
+- a lookup table that maps the `image version uid` to the CSP-specific image references
+
+The lookup table is uploaded to S3 and is used to identify the images that belong to a given `image version uid`.
+Measurements are uploaded to S3 and can be looked up for each cloud service provider and `image version uid`.
+
+## Image lookup table
+
+The image lookup table is a JSON file that maps the `image version uid` to the CSP-specific image references. It uses the `image version uid` as file name.
+
+```
+s3://<BUCKET-NAME>/images/<IMAGE-VERSION-UID>.json
+```
+
+```json
+{
+  "aws": {
+    "us-east-1": "ami-123",
+    "us-west-2": "ami-456",
+    "eu-west-1": "ami-789"
+  },
+  "azure": {
+    "cvm": "azure-cvm-123",
+    "trustedlaunch": "azure-trusted-launch-123"
+  },
+  "gcp": "gcp-image-123",
+  "qemu": "https://cdn.edgeless.systems/.../qemu.raw"
+}
+```
+
+- For AWS, the image lookup table contains the AMI IDs for each region
+- For Azure, the image lookup table contains the image IDs for the CVM and Trusted Launch images
+- For GCP, the image lookup table contains the image ID
+- For QEMU, the image lookup table contains a URL to the QEMU image
+
+This document is not signed and can be extended in the future to include more image references (e.g. if an image is replicated to a new AWS region).
+The same document can be used to identify old images that are no longer used and can be deleted for cost optimization.
+
+## Image measurements
+
+This RFC is not about the image measurements themselves, but about how they are stored and looked up.
+The format of the image measurements is described in the [secure software distribution RFC](secure-software-distribution.md).
+
+The image measurements are stored in a folder structure in S3 that is organized by CSP and `image version uid`.
+
+```
+s3://<BUCKET-NAME>/measurements/<CSP>/<IMAGE-VERSION-UID>/measurements.yaml
+s3://<BUCKET-NAME>/measurements/<CSP>/<IMAGE-VERSION-UID>/measurements.yaml.sig
+```
+
+## CLI image discovery
+
+The CLI needs to be able to discover the image references for a given `image version uid`.
+By default, the CLI will prefill the `image` field of the `constellation-conf.yaml` when `constellation config generate <CSP>` is run with a hardcoded `image version uid` (e.g. `v2.2.0-cc0de5c68d41f31dd0b284d574f137e0b0ad106b`).
```

Comment:
Why do we require the commit hash for release images?

I am thinking about a user wanting to upgrade their OS image. As I am understanding this right now they will have to lookup the version and hash. Not really a lot of work if you have the release page open already, but I also don't see why the commit is necessary for release versions as they should all be tagged.

Reply:
My thinking is that we really want to discourage manual management of this field. Down the line we want to discover what versions are available and compatible automatically in the CLI. If users see *just* a version field, they feel more inclined to modify it and break compatibility in the process. This also makes it just a little more streamlined for us to decide what commit an image was really built from - git tags can be set and moved later on and the release version string can be set to anything when running the pipeline as it is just a string input for the workflow.

I notice that you are not the first who feels uneasy about this so if the commit hash is super unpopular we can omit it.

Reply:
@Nirusu @derpsteb I think I got convinced after all to remove the hash from release image versions.